src/WebJobs.Script/Diagnostics/FileTraceWriter.cs

```diff
@@ -101,29 +86,38 @@ public override void Flush()
             }
 
             // write all lines in a single file operation
-            string contents = sb.ToString();
+            string content = sb.ToString();
             try
             {
-                lock (_syncLock)
-                {
-                    File.AppendAllText(_currentLogFileInfo.FullName, contents);
-                }
+                AppendToFile(_currentLogFileInfo, content);
             }
-            catch (DirectoryNotFoundException)
+            catch (Exception ex) when (ex is DirectoryNotFoundException || ex is FileNotFoundException)
             {
-                // need to handle cases where log file directories might
+                // need to handle cases where log files or directories might
                 // have been deleted from underneath us
-                Directory.CreateDirectory(_logFilePath);
-                lock (_syncLock)
-                {
-                    File.AppendAllText(_currentLogFileInfo.FullName, contents);
-                }
+                SetLogFile();
```

Comment:
Previously I was using AppendAllText which has "create if not exists" semantics. Which meant if we are using a file with a particular timestamp and it gets purged out from under us (e.g. because it is too old) we'll end up recreating that old file name. Now, if the file we were using gets deleted, we create a new one (or use an existing one if valid).